BPA/Ch.10/js_super_indran.stan

```diff
@@ -0,0 +1,254 @@
+// JS model using the superpopulation parameterization
+
+functions {
+  // These functions are derived from Section 12.3 of
+  // Stan Modeling Language User's Guide and Reference Manual
+
+  /**
+   * Return a integer value of first capture occasion
+   *
+   * @param y_i Integer array of capture history
+   *
+   * @return Integer value of first capture occasion
+   */
+  int first_capture(int[] y_i) {
+    for (k in 1:size(y_i))
+      if (y_i[k])
+        return k;
+    return 0;
+  }
+
+  /**
+   * Return a integer value of last capture occasion
+   *
+   * @param y_i Integer array of capture history
+   *
+   * @return Integer value of last capture occasion
+   */
+  int last_capture(int[] y_i) {
+    for (k_rev in 0:(size(y_i) - 1)) {
+      int k;
+      k <- size(y_i) - k_rev;
+      if (y_i[k])
+        return k;
+    }
+    return 0;
+  }
+
+  /**
+   * Return a matrix of uncaptured probability
+   *
+   * @param nind        Number of individuals
+   * @param n_occasions Number of capture occasions
+   * @param p           Detection probability for each individual
+   *                    and capture occasion
+   * @param phi         Survival probability for each individual
+   *                    and capture occasion
+   *
+   * @return Uncaptured probability matrix
+   */
+  matrix prob_uncaptured(int nind, int n_occasions,
+                         matrix p, matrix phi) {
+    matrix[nind, n_occasions] chi;
+
+    for (i in 1:nind) {
+      chi[i, n_occasions] <- 1.0;
+      for (t in 1:(n_occasions - 1)) {
+        int t_curr;
+        int t_next;
+
+        t_curr <- n_occasions - t;
+        t_next <- t_curr + 1;
+        chi[i, t_curr] <- (1.0 - phi[i, t_curr]) +
+                          phi[i, t_curr] *
+                          (1.0 - p[i, t_next]) *
+                          chi[i, t_next];
+      }
+    }
+    return chi;
+  }
+}
+
+data {
+  int<lower=0> M;                              // Augmented sample size
+  int<lower=0> n_occasions;                    // Number of capture occasions
+  int<lower=0,upper=1> y[M, n_occasions];      // Augmented capture-history
+}
+
+transformed data {
+  int<lower=0,upper=n_occasions> first[M];
+  int<lower=0,upper=n_occasions> last[M];
+
+  for (i in 1:M)
+    first[i] <- first_capture(y[i]);
+  for (i in 1:M)
+    last[i] <- last_capture(y[i]);
+}
+
+parameters {
+  real<lower=0,upper=1> mean_phi;             // Mean survival
+  real<lower=0,upper=1> mean_p;               // Mean capture
+  real<lower=0,upper=1> psi;                  // Inclusion probability
+  vector<lower=0>[n_occasions] beta;
+  vector[M] epsilon;
+  real<lower=0,upper=5> sigma;
```

Comment:
The declaration is changed to `real<lower=0,upper=20> sigma;`. The posterior 97.5% quantile of sigma^2 was 3.259.